services/construction.go

```diff
@@ -124,3 +145,713 @@ func (s *constructionAPIService) ConstructionSubmit(
+		request.Operations[2].Type == OpTransfer &&
+		request.Operations[2].Account.SubAccount != nil &&
+		request.Operations[2].Account.SubAccount.Address == SubAccountEscrow {
+		loggerCons.Debug("ConstructionPayloads: matched add escrow")
+		method = staking.MethodAddEscrow
+
+		if request.Operations[1].Account.Address != signWithAddr {
+			loggerCons.Error("ConstructionPayloads: add escrow from doesn't match signer",
+				"from", request.Operations[1].Account.Address,
+				"signer", signWithAddr,
+			)
+			return nil, ErrMalformedValue
+		}
+		amount, err := readOasisCurrencyNeg(request.Operations[1].Amount)
+		if err != nil {
+			loggerCons.Error("ConstructionPayloads: add escrow from amount",
+				"amount", request.Operations[1].Amount,
+				"err", err,
+			)
+			return nil, ErrMalformedValue
+		}
+
+		var escrowAccount staking.Address
+		if err = escrowAccount.UnmarshalText([]byte(request.Operations[2].Account.Address)); err != nil {
+			loggerCons.Error("ConstructionPayloads: add escrow account UnmarshalText",
+				"addr", request.Operations[2].Account.Address,
+				"err", err,
+			)
+		}
+		amount2, err := readOasisCurrency(request.Operations[2].Amount)
+		if err != nil {
+			loggerCons.Error("ConstructionPayloads: add escrow account amount",
+				"amount", request.Operations[2].Amount,
+				"err", err,
+			)
+			return nil, ErrMalformedValue
+		}
+		if amount.Cmp(amount2) != 0 {
+			loggerCons.Error("ConstructionPayloads: add escrow amounts differ",
+				"amount_from", amount,
+				"amount_to", amount2,
+				"err", err,
+			)
+			return nil, ErrMalformedValue
+		}
+
+		body = cbor.Marshal(staking.Escrow{
+			Account: escrowAccount,
+			Tokens:  *amount,
+		})
+	} else if len(request.Operations) == 2 &&
+		request.Operations[1].Type == OpTransfer &&
+		request.Operations[1].Account.SubAccount != nil &&
+		request.Operations[1].Account.SubAccount.Address == SubAccountEscrow {
+		loggerCons.Debug("ConstructionPayloads: matched reclaim escrow")
+		method = staking.MethodReclaimEscrow
+
+		var escrowAccount staking.Address
+		if err = escrowAccount.UnmarshalText([]byte(request.Operations[1].Account.Address)); err != nil {
+			loggerCons.Error("ConstructionPayloads: reclaim escrow from UnmarshalText",
+				"addr", request.Operations[1].Account.Address,
+				"err", err,
+			)
+		}
+		amount, err := readPoolShareNeg(request.Operations[1].Amount)
+		if err != nil {
+			loggerCons.Error("ConstructionPayloads: reclaim escrow from amount",
+				"amount", request.Operations[1].Amount,
+				"err", err,
+			)
+			return nil, ErrMalformedValue
+		}
+
+		body = cbor.Marshal(staking.ReclaimEscrow{
+			Account: escrowAccount,
+			Shares:  *amount,
+		})
+	} else {
+		loggerCons.Error("ConstructionPayloads: unmatched operations list",
+			"operations", request.Operations,
+		)
+		return nil, ErrNotImplemented
+	}
+
+	tx := transaction.Transaction{
+		Nonce: nonce,
+		Fee: &transaction.Fee{
+			Amount: *feeAmount,
+			Gas:    DefaultGas,
```

Comment:
How could this be changed to allow for gas estimation/specification? I guess the only part of this that is online is the metadata endpoint? So the preprocess thing would need to forward the operations, so that the metadata endpoint could construct a transaction for gas estimation. Then it would return an option with the gas part?

Even if we don't support gas estimation via Rosetta at this time, we should at least make it possible to specify the gas amount via an option instead of always forcing a default.

Reply:
> So the preprocess thing would need to forward the operations, so that the metadata endpoint could construct a transaction for gas estimation.

that's probably how you would do gas estimation. they are supposed to submit the operations in the preprocess step, so you'd naturally have access to them

> specify the gas amount via an option

ok